// my-api-utils/index.js
import { PrismaClient } from '@prisma/client';
import { PrismaMariaDb } from '@prisma/adapter-mariadb';
import {
    GetSecretValueCommand,
    SecretsManagerClient
} from '@aws-sdk/client-secrets-manager';

/* =========================
   RESPONSE UTILITIES
========================= */

const successResponse = (statusCode, data = null, headers = {}) => {
    const response = {
        statusCode,
        headers: {
            'Content-Type': 'application/json',
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Credentials': true,
            'Access-Control-Expose-Headers': 'Content-Range, X-Total-Count',
            ...headers
        }
    };

    if (data !== null) {
        response.body = JSON.stringify(data);
    }

    return response;
};

const errorResponse = (statusCode = 500, message = 'Internal Server Error') => ({
    statusCode,
    headers: {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Credentials': true
    },
    body: JSON.stringify({ message })
});

const parseBody = (body) =>
    typeof body === 'string' ? JSON.parse(body) : body;

/* =========================
   SECRETS MANAGER
========================= */

const getSecretValue = async (secretName) => {
    const config =
        process.env.APP_ENV === 'local'
            ? {
                region: process.env.AWS_REGION,
                credentials: {
                    accessKeyId: process.env.AWS_ACCESS_KEY_ID,
                    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY
                }
            }
            : {};

    const client = new SecretsManagerClient(config);

    const response = await client.send(
        new GetSecretValueCommand({ SecretId: secretName })
    );

    if (response.SecretString) {
        return JSON.parse(response.SecretString);
    }

    throw new Error('Secret not found or empty');
};

/* =========================
   DB CREDENTIALS
========================= */

const getDbCredentials = async () => {
    let credentials = {
        host: process.env.DB_HOST,
        username: process.env.DB_USERNAME,
        password: process.env.DB_PASSWORD,
        dbname: process.env.DB_DATABASE
    };

    if (process.env.DB_SECRET) {
        const secret = await getSecretValue(process.env.DB_SECRET);
        credentials = {
            host: secret.host,
            username: secret.username,
            password: secret.password,
            dbname: secret.dbname
        };
    }

    return credentials;
};

/* =========================
   PRISMA SINGLETON (Lambda)
========================= */

let prisma;

const getPrismaClient = async () => {
    if (prisma) return prisma;

    const db = await getDbCredentials();

    const adapter = new PrismaMariaDb({
        host: db.host,
        port: 3306,
        user: db.username,
        password: db.password,
        database: db.dbname,
        connectionLimit: 5
    });

    prisma = new PrismaClient({ adapter });

    return prisma;
};

/* =========================
   EXPORT
========================= */

export const utils = {
    parseBody,
    successResponse,
    errorResponse,
    getPrismaClient
};
